app/controllers/oaisys/pmh_controller.rb

```diff
@@ -2,25 +2,84 @@
 
 module Oaisys
   class PMHController < ApplicationController
+    EXPECTED_ARGUMENTS = {
+        identify: {
+            required: %w(),
+            optional: %w(),
+            exclusive: %w()
+        },
+        list_sets: {
+            required: %w(),
+            optional: %w(),
+            exclusive: %w(resumptionToken)
+        },
+        list_metadata_formats: {
+            required: %w(),
+            optional: %w(identifier),
+            exclusive: %w()
+        },
+        list_records: {
+            required: %w(metadataPrefix),
+            optional: %w(from until set),
+            exclusive: %w(resumptionToken)
+        },
+        get_record: {
+            required: %w(identifier metadataPrefix),
+            optional: %w(),
+            exclusive: %w()
+        },
+        list_identifiers: {
+            required: %w(metadataPrefix),
+            optional: %w(from until set),
+            exclusive: %w(resumptionToken)
+        },
+    }.freeze
     def bad_verb
     end
 
     def identify
+      expected_ags(__method__)
     end
 
     def list_sets
+      expected_ags(__method__)
     end
 
     def list_metadata_formats
+      expected_ags(__method__)
     end
 
     def list_records
+      expected_ags(__method__)
     end
 
     def get_record
+      expected_ags(__method__)
     end
 
     def list_identifiers
+      expected_ags(__method__)
+    end
+
+    private
+
+    def expected_ags (verb)
+      @verb = verb.to_s.camelize
+      @error_code = 'badArgument'
+      @error_message = 'The request includes illegal arguments or is missing required arguments.'
+      @oai_pmh_header = {
+          'xmlns': 'http://www.openarchives.org/OAI/2.0/',
+          'xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance',
+          'xsi:schemaLocation': 'http://www.openarchives.org/OAI/2.0/ http://www.openarchives.org/OAI/2.0/OAI-PMH.xsd'}
+
+      arguments = params.except('verb', 'controller', 'action').keys
+      expected_verb_arguments = EXPECTED_ARGUMENTS[verb].map { |_, arr| arr }.flatten
+      unexpected_arguments = !(arguments - expected_verb_arguments).empty?
+
+      required_verb_arguments = EXPECTED_ARGUMENTS[verb][:required]
+      missing_required_arguments = !(required_verb_arguments - arguments).empty?
+
+      render template: 'responses/error_response.xml.builder' if (unexpected_arguments || missing_required_arguments)
```

Comment:
For a lot of reasons, it's generally a bit of a code smell to set a bunch of different instance variables in an action and then use them in a view, even if it isn't a partial. I've found it makes future refactoring a lot harder, both because you then want to rewrite the instance variables in any piece of the view you extract into partials in the future, and because you can end up with "spooky action at a distance" when the controller-wide scoped instance variable gets touched by something that doesn't realize the instance variable is also used (invisibly) directly in views.

I'd suggest passing the verb, error code, and error message as locals instead. Another nice side-effect of doing it this way is that you can transparently (to the view) refactor the locals into methods down the road, which is another thing you can't do with directly-accessed instance variables.

Reply:
Oh, I finally found a good read that influenced this idea, https://thepugautomatic.com/2013/05/locals/

It raises another good point, which is that views will explode if you forget to pass/assign a local, but if they try to use an instance variable you forgot to assign, it just comes out as nil, which is harder to debug

Reply:
Makes sense, done